### PR TITLE
Cron job to check for and update images

### PR DESCRIPTION
**Pull request by kfox1111**
This patch adds a cron job to check images for updates and propose a PR with updated versions when one is found and a new version is available.

Running it, it already discovers 3 updates at the time of this writing.

It does have an issue with reformatting comments and removing white space. We may want to just go along with the reformatting the first time.

fixes: https://github.com/spiffe/helm-charts/issues/221

### CODE REVIEW

**Comment by kfox1111** [comment]
@marcofranssen Its failing shellcheck with:
./.github/scripts/update-tags.sh:38:85: note: Double quote to prevent globbing and word splitting. [SC2086]

But the non double quoted variable substitution is actually critical to its proper functioning.

**Comment by kfox1111** [comment]
Fixed

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:49]: There is a trivy action

https://github.com/aquasecurity/trivy-action

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:36]: Can we lock the action on the patch release.

```suggestion
        uses: actions/setup-go@v4.0.0
```

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:49]: The action runs trivy itself. We need to script the execution of it. Not sure how to do that with the upstream action.

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:36]: good idea.

File: .github/workflows/check-versions.yaml
Comment (faisal-memon) [reply .github/workflows/check-versions.yaml:49]: +1 for the trivy action if possible.

Comment (kfox1111) [reply .github/workflows/check-versions.yaml:49]: I looked into this some more. I don't think we can use the action:

1. It doesn't support just the install of trivy. its an action to run trivy.
2. We would no longer be able to just run the script locally on our own machines
3. We'd have to change things to be much much more complicated. It would have to do a dynamic matrix job that then runs all the trivy actions, then somehow tell it to not block on action failures, then run another job gathering up all the results of the trivy actions to do the building of the changes & pr.

So, might technically be possible to do, but feels undesirable.

Comment (kfox1111) [reply .github/workflows/check-versions.yaml:53]: I've been running this script manually since its not merged for a while now and its been very handy to be able to run it outside of gate.

File: .github/scripts/update-tags.sh
Comment (marcofranssen) [inline .github/scripts/update-tags.sh:49]:
```suggestion
    LATEST_VERSION=$(crane ls "${REGISTRY}/${REPOSITORY}" | grep "${FILTER}" | sort "${SORTFLAGS[@]}"| tail -n 1)
```

This might do the trick without disabling the linting. Not sure though in this usecase.

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:5]: Do we really want to increase the interval? Once a week seems sufficient to me.

File: .github/scripts/update-tags.sh
Comment (kfox1111) [reply .github/scripts/update-tags.sh:49]: No. the nonquoting is critical to its functioning. multiple args are passed through the variable and have to be handled separately by bash. The linter is trying to catch a common case of accidental misconfiguration but in this case, it really is intentional.

Comment (faisal-memon) [reply .github/scripts/update-tags.sh:49]: Can we put in a comment on why we have to disable the linter here?

Comment (kfox1111) [reply .github/scripts/update-tags.sh:50]: Done

File: .github/scripts/update-tags.sh
Comment (faisal-memon) [inline .github/scripts/update-tags.sh:59]: Is it better to put this in a file or at least have the statements on different lines so its easier to read?

**Comment by kfox1111** [comment]
Fixed.

File: .github/scripts/update-tags.sh
Comment (kfox1111) [reply .github/scripts/update-tags.sh:50]: > This might do the trick without disabling the linting. Not sure though in this usecase.

I tried it again just to be sure... the problem is the variable is a string with spaces, not an array, so the [@] form doesn't work.

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:5]: I'll revert it because weekly is still better then random, which we are doing now, and there is contention.

but, how do we know we have cve's without someone looking? This running is the thing that looks and lets us know. So if its weekly, we're only looking for cve fixes weekly. I'd rather know sooner.

The resources for the job are quite minimal, so its not a very high cost.

File: .github/scripts/update-tags.sh
Comment (kfox1111) [reply .github/scripts/update-tags.sh:50]: Will try.

Comment (kfox1111) [reply .github/scripts/update-tags.sh:50]: same issue. its trying to build one array argument out of a string intended to hold multiple:
sort: multi-character tab ‘ . -k1,1n -k2,2n -k3,3n’

I believe unquoted really is the correct usage.

You can test for yourself by setting spire-agent.waitForIt.image.tag  to latest-20220517
and running ./.github/scripts/update-tags.sh

Comment (edwbuck) [reply .github/scripts/update-tags.sh:50]: Both of these approaches work, and both work without shellcheck reporting an error, or needing quoting disabled.

```
SORTFLAGS=( "-k1,1n" "-k2,2n" "-k3,3n" )

sort "${SORTFLAGS[@]}" << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF

printf 'a, b, c, 1\na, c, c, 2\na, b, d, 3\n' | sort "${SORTFLAGS[@]}"
```

The idea that "${VARIABLE[@]}" expands to "-k1,1n -k2,2n -k3,3n" isn't true.  For arrays, expansion generally puts the double quotes around each element of an array expansion "-k1,1" "-k2,2" "-k3,3".

to illustrate, I use this bit of code to dump an array.  (It's from a shell library project I wrote, to manage function extensions in an "object_oriented" like manner)

```
function dump_object() {
        declare OBJECT_VAR=$1
        declare -n OBJECT_REF=$1

        SORTED_KEYS=()
        PREVIOUS_IFS=${IFS}
        while IFS= read -rd '' KEY; do
                SORTED_KEYS+=( "${KEY}" )
        done < <(printf '%s\0' "${!OBJECT_REF[@]}" | sort -z)
        for KEY in ${SORTED_KEYS[*]}; do
                echo ${OBJECT_VAR}[$KEY]=${OBJECT_REF[${KEY}]}
        done
        IFS=${PREVIOUS_IFS}
}
```

to use it one does
```
SORTFLAGS=( "-k1,1n" "-k2,2n" "-k3,3n" )
dump_object SORTFLAGS
```

to get 
```
SORTFLAGS[0]=-k1,1n
SORTFLAGS[1]=-k2,2n
SORTFLAGS[2]=-k3,3n
```

in the definition of SORTFLAGS above, the double quoting in the sort works perfectly.

```
#!/bin/bash

function dump_object() {
        declare OBJECT_VAR=$1
        declare -n OBJECT_REF=$1

        SORTED_KEYS=()
        PREVIOUS_IFS=${IFS}
        while IFS= read -rd '' KEY; do
                SORTED_KEYS+=( "${KEY}" )
        done < <(printf '%s\0' "${!OBJECT_REF[@]}" | sort -z)
        for KEY in ${SORTED_KEYS[*]}; do
                echo ${OBJECT_VAR}[$KEY]=${OBJECT_REF[${KEY}]}
        done
        IFS=${PREVIOUS_IFS}
}

SORTFLAGS=( "-k1,1n" "-k2,2n" "-k3,3n" )

dump_object SORTFLAGS

sort "${SORTFLAGS[@]}" << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF
```

but if I do
```
BAD_SORTFLAGS=( "-k1,1n -k2,2n -k3,3n" )

dump_object BAD_SORTFLAGS
```
which shows as output
```
BAD_SORTFLAGS[0]=-k1,1n -k2,2n -k3,3n
```

Then running
```
sort "${BAD_SORTFLAGS[@]}" << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF
```
or
```
printf 'a, b, c, 1\na, c, c, 2\na, b, d, 3\n' | sort "${SORTFLAGS[@]}"
```
doesn't work but the unquoted versions
```
sort ${BAD_SORTFLAGS[@]} << 'EOF'
a, b, c, 1
a, c, c, 2
a, b, d, 3
EOF
```
and
```
printf 'a, b, c, 1\na, c, c, 2\na, b, d, 3\n' | sort ${SORTFLAGS[@]}
```
do work.

Odds are you are misidentifying the root cause.  It's not because the quotes around ${SORTFLAGS[@]} are combining the parameters, it's because you only set one value in ${SORTFLAGS[@]} that has spaces separating the text within that single value.

Please never disable the quoting rules in shellcheck, of all the rules, it's the most important one.

File: .github/scripts/update-tags.sh
Comment (edwbuck) [inline .github/scripts/update-tags.sh:1]: Need to check with the object_dump the actual values within SORTFLAGS, which are probably not set the way they are thought to be set, which is why the removed quoting works when the proper quoting doesn't.

**Review by edwbuck — CHANGES_REQUESTED** [review]

File: .github/scripts/update-tags.sh
Comment (kfox1111) [reply .github/scripts/update-tags.sh:50]: The source of the string is not a bash array though, but a json document. How do you suggest we pull it into a bash array?

Comment (kfox1111) [reply .github/scripts/update-tags.sh:50]: Its defined in the json config like:
https://github.com/spiffe/helm-charts/pull/249/files#diff-4c6c57c7cd483f8116b836fc3762a95f0f347b5f1f5e4cb67002f247d50ff14bR24

Then pulled into bash here:
https://github.com/spiffe/helm-charts/pull/249/files#diff-e800c023ce454983c68686a85f2ea49f6114c2cd433a239eb9637c84ae3ddeefR42

There could be a better way to structure that, but no idea what that should look like

**Comment by marcofranssen** [comment]
Thinking again on this, isn't it better if the upstream project checks for vulnerabilities in the images? Maybe contribute it there on the images we are using. 

In the end it doesn't help us to detect the vulnerability if the upstream doesn't make a release. I think we should just check for new version like dependabot does. Checking for vulnerabilities seems to be something that shouldn't be part of this helm-charts repo, but should be handled in the upstream projects we depend on (hence why I advised to use chainguard images), then we can have a job that checks for new releases.

IMHO we should close this PR as image scanning shouldn't be our responsibility as in the end it doesn't give us value in the sense of having patches for those things we detect.

**Comment by faisal-memon** [comment]
> Thinking again on this, isn't it better if the upstream project checks for vulnerabilities in the images? Maybe contribute it there on the images we are using.
> 
> In the end it doesn't help us to detect the vulnerability if the upstream doesn't make a release. I think we should just check for new version like dependabot does. Checking for vulnerabilities seems to be something that shouldn't be part of this helm-charts repo, but should be handled in the upstream projects we depend on (hence why I advised to use chainguard images), then we can have a job that checks for new releases.
> 
> IMHO we should close this PR as image scanning shouldn't be our responsibility as in the end it doesn't give us value in the sense of having patches for those things we detect.

I agree with the above. In the end if we are just staying up to date on the latest versions then we are just as good. The code to do that should be much simpler. I spent a good chunk of time reviewing but could not wrap my head around the scripts. I wish I had your scripting skills @kfox1111. 

With so many other pr's to review i'm not sure how much more time i can spend on this one.

**Comment by kfox1111** [comment]
> Thinking again on this, isn't it better if the upstream project checks for vulnerabilities in the images? Maybe contribute it there on the images we are using.
> 
> In the end it doesn't help us to detect the vulnerability if the upstream doesn't make a release. I think we should just check for new version like dependabot does. Checking for vulnerabilities seems to be something that shouldn't be part of this helm-charts repo, but should be handled in the upstream projects we depend on (hence why I advised to use chainguard images), then we can have a job that checks for new releases.
> 
> IMHO we should close this PR as image scanning shouldn't be our responsibility as in the end it doesn't give us value in the sense of having patches for those things we detect, those patches will only arrive if the upstream detects them, hence why I advise to contribute this over there in the upstream projects.

Chainguard releases daily. So if we do what you suggest, we will be patching our charts, daily. The container scanner part of the pr does just one thing. Squash patching version when no cve's are fixed so we're not having to cut releases constantly to deal with the fact that chainguard releases daily and we can't easily tell which ones are important to update (actual cve's fixed) vs just a new daily image was built. We could consider not doing the trivy stuff if we stopped using chainguard as they are the only images doing such constant releases.

**Comment by kfox1111** [comment]
Are we trying to throw the baby out with the bathwater? When I joined up, I think the only one that knew how the .github tests worked was Marco. It was ok. Things were better with the tests then without them.

Perhaps this is in that category. We are only better off with this in place then without it. That not everyone knows all the gory details of how it works in its entirety, is not great, but we're still better off with it IMO.

Based on how complex it became in bash, its probably better to rewrite this in something more well known such as python. I propose though that we can merge this, as is, and when other, more pressing issues are out of the way, we can do that then. I'm afraid we're letting the perfect stand in the way of the good enough here.

**Comment by kfox1111** [comment]
Ok. With Edwin's help, we got it converted over to read into a bash array. This fixes the shellcheck issue so we don't have to squash it. Thanks Edwin!

**Review by edwbuck — APPROVED** [review]
Worked with Kevin to get all BASH passing shellcheck.

Led to actually documenting the different flags in the input YAML as different flags (previously there were not an array, but a string).

Led to changes to pull each array item into a bash array,

Led to the ability to properly quote that array on expansion.

**Comment by edwbuck** [comment]
@faisal-memon @marcofranssen @mrsabath Can we get a second review here?  This is nearly ready to merge.

**Comment by faisal-memon** [comment]
> @faisal-memon @marcofranssen @mrsabath Can we get a second review here?  This is nearly ready to merge.

@edwbuck Any thoughts on the concerns posted earlier and the alternatives presented?

**Comment by kfox1111** [comment]
Anything left to do?

**Comment by faisal-memon** [comment]
@kfox1111 I appreciate proactively focusing on security. However there are alot of concerns presented above with this pr. What value is the trivy scanning providing vs just keep versions up to date?

**Comment by kfox1111** [comment]
> @kfox1111 I appreciate proactively focusing on security. However there are alot of concerns presented above with this pr. What value is the trivy scanning providing vs just keeping versions up to date?

I already explained that here:
https://github.com/spiffe/helm-charts/pull/249#issuecomment-1582629766

Please respond to that along with this: 
https://github.com/spiffe/helm-charts/pull/249#issuecomment-1582770047

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:54]: Shouldn't we use a requirements.txt so we can configure dependabot to keep these deps updated? This is likely to break if those deps change.

https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#package-ecosystem

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:41]: Lock on the patch release like we do with other actions.

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:36]: Why do we need Go in this pipeline? Doesn't seem like we need to compile any Go here.

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [inline .github/workflows/check-versions.yaml:47]: Since we are doing a python script anyhow, why not resolve this in python? so we don't need to do the yq part.

File: .github/tests/images.json
Comment (marcofranssen) [inline .github/tests/images.json:1]: Some newly added images are missing here.

File: .github/tests/images.json
Comment (marcofranssen) [inline .github/tests/images.json:35]: We switched to use the semver tag here. So this also abstracts away the need of daily patching.

File: .github/tests/images.json
Comment (marcofranssen) [inline .github/tests/images.json:41]: We switched to use the semver tag here. So this also abstracts away the need of daily patching.

File: .github/tests/images.json
Comment (marcofranssen) [inline .github/tests/images.json:5]: We switched to use the semver tag here. So this also abstracts away the need of daily patching.

**Review by marcofranssen — COMMENTED** [review]
I gave it one more review, still think this is a lot of hassle to keep things patched, while we actually should rely on the upstream projects to provide those patches.

Also see the inline comments on the bash image. The chart now uses the semver version which is not releasing daily. It releases whenever there is a new patch for bash, meaning only when a CVE is resolved.

So I still heavily doubt we need all of this stuff in our charts repo. In the end we can't patch anyhow if there is no patch released.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
I gave it one more review, still think this is a lot of hassle to keep things patched, while we actually should rely on the upstream projects to provide those patches.

Also see the inline comments on the bash image. The chart now uses the semver version which is not releasing daily. It releases whenever there is a new patch for bash, meaning only when a CVE is resolved.

So I still heavily doubt we need all of this stuff in our charts repo. In the end we can't patch anyhow if there is no patch released.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
I gave it one more review, still think this is a lot of hassle to keep things patched, while we actually should rely on the upstream projects to provide those patches.

Also see the inline comments on the bash image. The chart now uses the semver version which is not releasing daily. It releases whenever there is a new patch for bash, meaning only when a CVE is resolved.

So I still heavily doubt we need all of this stuff in our charts repo. In the end we can't patch anyhow if there is no patch released.

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:54]: We could. Where should we put the file so that dependabot could find them though?

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:41]: For crane, they only have major.minor

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:36]: left over from yq I think. Will remove.

Comment (kfox1111) [reply .github/workflows/check-versions.yaml:36]: oh. yq is still used.

File: .github/workflows/check-versions.yaml
Comment (kfox1111) [reply .github/workflows/check-versions.yaml:47]: See comment about maybe rewriting this whole feature to be all python later. It started life in bash/yq and then gained python just to do the editing without messing up the comments.

File: .github/tests/images.json
Comment (kfox1111) [reply .github/tests/images.json:1]: Not sure which ones you mean? the kubectl ones don't have versions pinned.

File: .github/tests/images.json
Comment (kfox1111) [reply .github/tests/images.json:35]: can switch to semver for bash. waitForIt has no non latest tags.

Comment (kfox1111) [reply .github/tests/images.json:35]: Have a look at https://github.com/spiffe/helm-charts/pull/356#pullrequestreview-1484828752
I think we probably want to stick with latest variants because they are immutable.

File: .github/tests/images.json
Comment (kfox1111) [reply .github/tests/images.json:41]: same

File: .github/tests/images.json
Comment (kfox1111) [reply .github/tests/images.json:5]: same

**Comment by kfox1111** [comment]
> I gave it one more review, still think this is a lot of hassle to keep things patched, while we actually should rely on the upstream projects to provide those patches.

How do we do that without actually looking for updated versions? If we can find a way to use versioned tags with chainguard we could drop the need for the trivy scanner from the script. That doesn't negate the need for the rest of the script?

> Also see the inline comments on the bash image. The chart now uses the semver version which is not releasing daily. It releases whenever there is a new patch for bash, meaning only when a CVE is resolved.

see https://github.com/spiffe/helm-charts/pull/356#pullrequestreview-1484828752

> So I still heavily doubt we need all of this stuff in our charts repo. In the end we can't patch anyhow if there is no patch released.

How do we know we need to patch without this script? Are you looking? I see, just manually running the script, there is another image needing an update. I don't know how long its been needed as its not running periodically and I"m not checking daily.

**Comment by faisal-memon** [comment]
> Chainguard releases daily. So if we do what you suggest, we will be patching our charts, daily. 

We can just run the version check cron once a week, so we wont be patching daily.

> So I still heavily doubt we need all of this stuff in our charts repo. In the end we can't patch anyhow if there is no patch released.

Agree. 

> How do we know we need to patch without this script? Are you looking?

Same as we do with dependabot. Run a simple version checker once a week.

**Comment by kfox1111** [comment]
> Same as we do with dependabot. Run a simple version checker once a week.

How is that not this script, minus the few lines that do the trivy run?

**Comment by faisal-memon** [comment]
> How is that not this script, minus the few lines that do the trivy run?

Its the trivy part that we are both questioning. New versions of images with critical bug fixes but no cve's will be filtered out by it. So we will still need to manually check for new images anyways.

**Comment by kfox1111** [comment]
> > How is that not this script, minus the few lines that do the trivy run?
> 
> Its the trivy part that we are both questioning. New versions of images with critical bug fixes but no cve's will be filtered out by it. So we will still need to manually check for new images anyways.

Ok, it sounds like we may be closer then I was thinking. I was confused when @marcofranssen  said:
> IMHO we should close this PR as image scanning shouldn't be our responsibility

And you seemed to agree with him.

The trivy part is just a few lines. I'm ok getting rid of it entirely, IF,
waitForIt actually cuts a versioned release and we can figure out how to get immutable tags from chainguard. Until then, the script without trivy would basically cause a patch to be created basically every run, which would cause more work for us, and more work for end users if it didn't actually fix anything.

We also could add a flag to the config to make the cve scanner optional for all images that have a release version if that would help too?

**Comment by kfox1111** [comment]
chainguard image issues:
https://github.com/chainguard-images/images/issues/881
and
https://github.com/chainguard-images/images/issues/880

881 may make 880 moot if they really only support immutable images with the latest-xxx tags.

**Comment by kfox1111** [comment]
Sounds like we need to update the logic to always use :latest@sha256xxxx as per https://github.com/spiffe/helm-charts/issues/442

**Comment by kfox1111** [comment]
Updated to deal with https://github.com/spiffe/helm-charts/issues/442

**Review by faisal-memon — APPROVED** [review]
Thanks @kfox1111 for creating this and removing the trivy scanner part.

File: .github/workflows/check-versions.yaml
Comment (marcofranssen) [reply .github/workflows/check-versions.yaml:50]: You put it in the same folder as the python script.

See:

https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file

I think it should be using pip as package manager and you can define the path to the folder containing the python code and requirements txt.

File: .github/tests/images.json
Comment (marcofranssen) [reply .github/tests/images.json]: Oh found the answer further down in this thread.

https://github.com/spiffe/helm-charts/pull/249#issuecomment-1674018842

File: .github/tests/images.json
Comment (kfox1111) [reply .github/tests/images.json:1]: fsGroup was added since, so I just added it here.

**Review by marcofranssen — APPROVED** [review]
LGTM 🚀

**Pull request merged**
